### PR TITLE
Insert plan changes

### PR DESCRIPTION
**Pull request by pervazea**
Fixes for #1197.
Tests are in separate PR in scripts/testing/junit

### CODE REVIEW

**Comment by pmenon** [comment]
@pervazea If you don't mind, I'm changing some insert stuff for multithreading. I'll take a look at this too.

**Comment by pervazea** [comment]
Updated in accordance with review comments.

File: src/include/planner/insert_plan.h [hunk 1]
```diff
@@ -29,6 +29,26 @@ class InsertStatement;
 }
 
 namespace planner {
+
+// mapping from schema columns to insert columns
+struct SchemaColsToInsertCols {
```
Comment (pmenon) [inline src/include/planner/insert_plan.h]: If this isn't needed in the header, can you move to implementation?

File: src/include/planner/insert_plan.h [hunk 2]
```diff
@@ -57,18 +77,73 @@ class InsertPlan : public AbstractPlan {
...
+   */
+  InsertPlan(storage::DataTable *table,
+	     const std::vector<std::string> *columns,
+             const std::vector<std::vector<std::unique_ptr<expression::AbstractExpression>>> *insert_values);
 
   // Get a varlen pool - will construct the pool only if needed
   type::AbstractPool *GetPlanPool();
 
-  PlanNodeType GetPlanNodeType() const override { return PlanNodeType::INSERT; }
+  PlanNodeType GetPlanNodeType() const override {
```
Comment (pmenon) [inline src/include/planner/insert_plan.h]: Format.

File: src/include/planner/insert_plan.h [hunk 3]
```diff
@@ -126,6 +201,11 @@ class InsertPlan : public AbstractPlan {
   // Values
   std::vector<type::Value> values_;
 
+  // mapping from schema columns to vector of insert columns
+  std::vector<struct SchemaColsToInsertCols> schema_to_insert_;
```
Comment (pmenon) [inline src/include/planner/insert_plan.h]: No need for `struct` keyword.

File: src/planner/insert_plan.cpp [hunk 4]
```diff
@@ -28,84 +28,180 @@ InsertPlan::InsertPlan(storage::DataTable *table,
...
+	  // get index into values
+	  auto val_idx = schema_to_insert_[idx].val_idx;
+	  auto &exp = values[val_idx];
+	  auto exp_ptr = exp.get();
+	  bool ret_bool = ProcessValueExpr(exp_ptr, idx);
+	  if (ret_bool) {
+	    is_prepared_stmt = true;
+	  }
+	}
+	else {
```
Comment (pmenon) [inline src/planner/insert_plan.cpp]: Else on new line.

File: src/include/planner/insert_plan.h [hunk 5]
```diff
@@ -57,18 +77,73 @@ class InsertPlan : public AbstractPlan {
...
+   * Process a single expression to be inserted.
+   *
+   * @param[in] expr       insert expression
+   * @param[in] schema_idx index into schema columns, where the expr
+   *                       will be inserted.
+   * @return  true if values imply a prepared statement
+   *          false if all values are constants. This does not rule
+   *             out the insert being a prepared statement.
+   */
+  bool ProcessValueExpr(expression::AbstractExpression *expr,
```
Comment (pmenon) [inline src/include/planner/insert_plan.h]: Mark all these functions are private if they're not needed externally.

**Comment by pervazea** [comment]
@pmenon _Do we have tests to test this new logic?_
See script/testing/junit directory

These are not run automatically yet, but the intention is that they will be integrated in and run by Jenkins in some suitable way.

**Pull request merged**